packages/ui/src/Shared/DndFileChooser.tsx

```diff
@@ -0,0 +1,312 @@
+   */
+  i18nNoFileSelectedMessage: string;
+
+  /**
+   * The localized text for the label that appears before the selected file information.
+   */
+  i18nSelectedFileLabel: string;
+
+  /**
+   * A general, localized message for when a file upload failed. This message will be shown
+   * along with an error icon and appears next to the selected file message. If this property
+   * is set then `i18nUploadSuccessMessage` should not be set.
+   */
+  i18nUploadFailedMessage?: string;
+
+  /**
+   * A general, localized message for when a file upload was successful. This message will be shown
+   * along with an OK icon and appear next to the selected file message. If this property
+   * is set then `i18nUploadFailedMessage` should not be set.
+   */
+  i18nUploadSuccessMessage?: string;
+
+  /**
+   * The number of milliseconds that toast notifications will be displayed. Defaults to `4000`.
+   */
+  notificationTimerDelay?: number;
+
+  /**
+   * Obtains the localized text (may include HTML tags) that appears when the file upload was rejected. This
+   * will occur when a DnD of a file with the wrong extension is dropped. This message is presented
+   * as a timed toast notification.
+   */
+  onUploadRejected(fileName: string): string;
+
+  /**
+   * Callback for when one or more file uploads have been accepted. Caller should handler processing of the files.
+   */
+  onUploadAccepted(file: File[]): void;
+}
+
+/**
+ * A type used to identify files that were rejected. The key is the file name.
+ */
+interface INotificationType {
+  key: string;
+  message: string;
+}
+
+/**
+ * The state properties of the `DndFileChooser`.
+ */
+export interface IDndFileChooserState {
+  /**
+   * Indicates if multiple files can be added. Defaults to `false`.
+   */
+  allowMultiple: boolean;
+
+  /**
+   * The files that were uploaded successfully. Defaults to an empty array.
+   */
+  files: File[];
+
+  /**
+   * The error notifications for rejected files. After a toast is displayed for a notification it is removed
+   * from the array.
+   */
+  notifications: INotificationType[];
+
+  /**
+   * The number of milliseconds that toast notifications will be displayed. Defaults to `4000`.
+   */
+  notificationTimerDelay: number;
+}
+
+/**
+ * A component that can be used to upload files. Files can be uploaded by clicking in the drop zone or by
+ * drag and dropping files into the drop zone.
+ */
+export class DndFileChooser extends React.Component<
+  IDndFileChooserProps,
+  IDndFileChooserState
+> {
+  public constructor(props: IDndFileChooserProps) {
+    super(props);
+
+    // set up initial state
+    this.state = {
+      allowMultiple: this.props.allowMultiple
+        ? this.props.allowMultiple
+        : false,
+      files: [],
+      notificationTimerDelay: this.props.notificationTimerDelay
+        ? this.props.notificationTimerDelay
+        : 4000,
+      notifications: [],
+    };
```

Comment:
You are storing as state something that comes from the props, but you are not dealing with prop changes. This is a common anti-pattern. 

Check `Principle 1: Don’t Stop the Data Flow` here for more details https://overreacted.io/writing-resilient-components/#dont-stop-the-data-flow-in-rendering

Reply:
I was thinking those 2 properties (allowMultiple, notificationTimerDelay) would not change. I was trying to add default values for those properties so that the parent didn't have to set them if they didn't want to. I can see your point where this might be confusing or maybe not behave the way someone might assume it does. Is there a better way to do this?

Reply:
https://reactjs.org/docs/typechecking-with-proptypes.html#default-prop-values

[This article](https://medium.com/@martin_hotell/react-typescript-and-defaultprops-dilemma-ca7f81c661c7) will give you an idea of the pitfalls of `defaultProps` with TS, but I wouldn't bother going that route, I'd just set the `defaultProps` static property and call it a day.

Reply:
I'll update and use the static `defaultProps` property.